Source/DafnyCore/Verifier/Translator.cs

```diff
@@ -3851,6 +3851,122 @@ public enum StmtType { NONE, ASSERT, ASSUME };
+                return b1.CompareTo(b2) != 0;
+              }
+              if (e0 is LiteralExpr { Value: bool bb1 } && e1 is LiteralExpr { Value: bool bb2 }) {
+                return bb1 != bb2;
+              }
+              if (!truth &&
+                  e0.Type.NormalizeExpand().AsIndDatatype is IndDatatypeDecl
+                  && e1.Type.NormalizeExpand().AsIndDatatype is IndDatatypeDecl
+                  && Printer.ExprToString(e0) is var e0str
+                  && Printer.ExprToString(e1) is var e1str
+                  && (e1str.StartsWith(e0str + ".") || (e0str.StartsWith(e1str + ".")))) {
+                return true;
+              }
+              return false;
+            }
+          case BinaryExpr.Opcode.Lt: {
+              if (!truth && Printer.ExprToString(e0) == Printer.ExprToString(e1)) { // Obvious case when A < B is wrong: B == A
+                return true;
+              }
+              // Obvious case when A < B is always true, when A and B are bigintegers
+              if (e0 is LiteralExpr { Value: BigInteger b1 } && e1 is LiteralExpr { Value: BigInteger b2 }) {
+                return truth == (b1.CompareTo(b2) < 0);
+              }
+              return false;
+            }
+          case BinaryExpr.Opcode.Gt: {
+              // Obvious case when A > B is wrong: B == A
+              if (!truth && Printer.ExprToString(e0) == Printer.ExprToString(e1)) {
+                return true;
+              }
+              // Obvious case when A > B is always true, when A and B are bigintegers
+              if (e0 is LiteralExpr { Value: BigInteger b1 } lit1 && e1 is LiteralExpr { Value: BigInteger b2 }) {
+                return truth == (b1.CompareTo(b2) > 0);
+              }
+              return false;
+            }
+          case BinaryExpr.Opcode.Neq: {
+              // Obvious case when A != B is wrong: B == A
+              if (!truth && Printer.ExprToString(e0) == Printer.ExprToString(e1)) {
+                return true;
+              }
+              // Obvious case when A > B is always true, when A and B are bigintegers
+              if (e0 is LiteralExpr { Value: var value1 } && e1 is LiteralExpr { Value: var value2 }) {
+                if (value1 is BigInteger b1 && value2 is BigInteger b2) {
+                  return truth == (b1.CompareTo(b2) != 0);
+                } else if (value1 is bool bb1 && value2 is bool bb2) {
+                  return truth == (bb1 != bb2);
+                }
+              }
+              if (truth
+                  && e0.Type.NormalizeExpand().AsIndDatatype is IndDatatypeDecl
+                  && e1.Type.NormalizeExpand().AsIndDatatype is IndDatatypeDecl
+                  && Printer.ExprToString(e0) is var e0str
+                  && Printer.ExprToString(e1) is var e1str
+                  && (e1str.StartsWith(e0str + ".") || (e0str.StartsWith(e1str + ".")))) {
```

Comment:
Why do you have to invoke the printer as part of this static analysis?

Reply:
It makes it easy to reason about expressions equality. It is slower than extracting the original string, but it's more robust. It's slower than if we had an up-to-date and well maintained Equals() function on expressions, but this was never implemented. This could be done in the future, and I keep the scope of this PR as small as possible,
Printers are linear and cost a linear amount of memory so I don't think this is much of a concern.

Reply:
I don't know :/ I'd rather just remove this feature than depending on the printer for static analysis.

Reply:
I meant removing this specific part, not the whole feature.  It's an approximation anyway, so making it a bit less precise isn't an issue, is it?

Reply:
Besides, isn't the check wrong?

```
datatype T = A | B {
  const ths := this
}

method M(t: T) {
  assert t.ths == t;
  assume t.ths != t; // Should be flagged
}
```

I really don't think relying on pretty-printing is the right way to go.

Reply:
> I meant removing this specific part, not the whole feature. It's an approximation anyway, so making it a bit less precise isn't an issue, is it?

Making it a bit less precise just wouldn't have caught multiple development issues I have witness in the past year, such as `i < i`, `SomeFunction(a, b) != SomeFunction(a, b)`, or recently @dijkstracula 's  `a[i] != a[i]`.
I would feel very bad after this PR if someone had such an issue of this type again, and I told him, "I'm sorry, your code could have been flagged if we just hadn't removed this small part from the PR"

> I really don't think relying on pretty-printing is the right way to go.

I get your point that we don't want to ship half-baked features. Every time we issue a warning, users will start relying on these warnings in their everyday's development process, and might get mad if some warnings don't show up where they think they should.

Now, I never claimed relying on pretty printing is the "right" way to go. I claim I claim it's a cheap, brutal and efficient way to get the job to start.
Plus, since it's not clever, users are quickly going to realize it - and won't be surprised in your example that the warning isn't issued.
Take as example that no one entirely on the null-pointer analysis of C#, or its warning that a condition is always true/false because it's never complete, there are always ways to create a condition that is trivially false and that is not flagged, for example.
So, especially because we are generating warnings and not errors, we don't have any reason to believe that users will have high expectations of these warnings

Defining a scope for quick static analysis can be done later - not by me, I'm working on proof locality, not static analysis - and will be another topic. I'm not sure who will want to work on it, but if that means we skip these already useful warnings for the next 12 months, I would just strongly oppose against removing this feature in this PR.

Reply:
But just to be clear: that test is wrong, right?

Reply:
Oh I completely oversaw the test itself. My head too much trying to describe the benefits of implementing this PR cheaply :-)

Yes, you're right. I did not think of inner fields destructions. I made the test robust by ensuring that the indirection is a datatype destructor. I also added a test to capture your behavior.

Thank you very much for this contribution !